### PR TITLE
Use saturating_add in merge_results

### PR DESCRIPTION
**Pull request by lu-zero**

### CODE REVIEW

**Comment by calixteman** [comment]
I don't think it's a good idea.
Counters are u64, so in normal conditions it's impossible to reach the max... and so it's very likely a good way to hide a real problems and maybe show inaccurate results.

**Comment by marco-c** [comment]
@calixteman between the two options (overflowing and risking a 0 counter and saturating), the saturating option seems better to me.
We can print a warning in case the add would overflow to avoid "hiding" the problem.

**Comment by marco-c** [comment]
@lu-zero as part of this, could you also revert 155c73f05c67f173c75946b07e22b9d53f65d647 (thus fixing #525).

**Comment by calixteman** [comment]
I'm fine with having an error at least.
I wonder if -1 is more or less equivalent `to counter is 0` or `no counter`.
I'd say have a look on where are these bad counters and then choose on of the above options.
And if none of the above options is correct then use `saturating_add` (with a TODO) is fine.
But we really need to investigate either on llvm's side or rustc's one where is the problem.

**Comment by lu-zero** [comment]
To reproduce the problem you can run gcovr on the gcovr tests :) Let me know how you'd like to proceed.

**Comment by marco-c** [comment]
@lu-zero let's:
1. print a warning if there could be an overflow;
2. don't overflow but saturate.

Also get rid of the workaround from 155c73f05c67f173c75946b07e22b9d53f65d647 which should no longer be necessary.

File: src/lib.rs [hunk 1]
```diff
@@ -65,7 +65,11 @@ pub fn merge_results(result: &mut CovResult, result2: CovResult) {
     for (&line_no, &execution_count) in &result2.lines {
         match result.lines.entry(line_no) {
             btree_map::Entry::Occupied(c) => {
-                *c.into_mut() += execution_count;
+                if c.get().checked_add(execution_count).is_none() {
+                    warn!("Execution count overflow detected.");
```
Comment (calixteman) [inline src/lib.rs:69]: warn only one time.

File: src/lib.rs [hunk 2]
```diff
@@ -65,7 +65,11 @@ pub fn merge_results(result: &mut CovResult, result2: CovResult) {
...
+                }
+
+                *c.into_mut() = c.get().saturating_add(execution_count);
```
Comment (calixteman) [inline src/lib.rs:72]: if add is correct from the above check then no need to call saturating_add.

File: Cargo.toml [hunk 3]
```diff
@@ -66,7 +66,7 @@ flate2 = "^1.0"
 lazy_static = "^1.3"
 
 [profile.release]
-lto = true
+lto = "thin"
```
Comment (calixteman) [inline Cargo.toml:69]: why this change ?

Comment (lu-zero) [reply Cargo.toml:69]: Marco asked for a patch about it, it ended up in this set, I can make a separate PR for it.

File: src/lib.rs [hunk 4]
```diff
@@ -61,11 +61,17 @@ use std::path::PathBuf;
...
-pub fn merge_results(result: &mut CovResult, result2: CovResult) {
+pub fn merge_results(result: &mut CovResult, result2: CovResult) -> bool {
+    let mut warn_overflow = false;
     for (&line_no, &execution_count) in &result2.lines {
         match result.lines.entry(line_no) {
             btree_map::Entry::Occupied(c) => {
-                *c.into_mut() += execution_count;
+                let v = c.get().checked_add(execution_count).unwrap_or_else(|| {
+                    warn_overflow = true;
+                    c.get().saturating_add(execution_count)
```
Comment (calixteman) [inline src/lib.rs:71]: in this case just return u64::max: no need to call saturating_add

**Review by calixteman — APPROVED** [review]
